### PR TITLE
Issue #71 KPM account

### PR DESCRIPTION
**Pull request by wwjbatista**

### CODE REVIEW

File: kpm/lib/kpm/account.rb [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+      if account_id === :export.to_s
+        puts "\e[91;1mNeed to specify an account id\e[0m\n\n"
+        Dir.rmdir(TMP_DIR)
+        return
+      end
+
+      uri = URI("#{@killbill_url}/#{KILLBILL_API_VERSION}/kb/export/#{account_id}")
+      export_file = TMP_DIR + File::SEPARATOR + 'kbdump'
+
+      request = Net::HTTP::Get.new(uri.request_uri)
```
Comment (pierre) [inline kpm/lib/kpm/account.rb]: Could you use the Ruby client instead? See https://github.com/killbill/killbill-client-ruby.

Note: it doesn't look like the export endpoint is implemented yet in the client, so it would need to be added first.

File: kpm/lib/kpm/account.rb [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+          @config = YAML::load_file(config_file)
+        end
+      end
+
+    end
+
+    def export_data(account_id = nil)
+
+      if account_id === :export.to_s
+        puts "\e[91;1mNeed to specify an account id\e[0m\n\n"
```
Comment (pierre) [inline kpm/lib/kpm/account.rb]: Use a `Logger` instead, so we can easily toggle `DEBUG` if need be.

File: kpm/lib/kpm/account.rb [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        Dir.rmdir(TMP_DIR)
```
Comment (pierre) [inline kpm/lib/kpm/account.rb]: Shouldn't this be done automatically (`mktmpdir`) behind the scenes?

Always worried about `rm` calls in such scripts that are potentially run on production systems... 😅

File: kpm/lib/kpm/account.rb [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+      end
+
+      if not database.nil?
+
+        Database.set_configuration(database[0], database[1], database[2])
+
+      end
+    end
+
+    def set_config(config_file = nil)
```
Comment (pierre) [inline kpm/lib/kpm/account.rb]: Can you separate such helpers at the bottom and mark them as `private`? It would help the overall reading and distinguish the _public APIs_.

File: kpm/lib/kpm/account.rb [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+
+      open (export_file), 'w' do |io|
+        #io.write response.body
+        table_name = nil
+        cols_names = nil
+        response.body.split("\n").each do |line|
+          words = line.strip.split(" ")
+          clean_line = line
+
+          if /--/.match(words[0])
```
Comment (pierre) [inline kpm/lib/kpm/account.rb]: Could you extract small utility methods out of the main flow and write unit tests for it? This will help with maintenance in the future.

File: kpm/lib/kpm/tasks.rb [hunk 6]
```diff
@@ -432,6 +432,97 @@ def system
...
+                      :type    => :string,
+                      :default => nil,
+                      :desc    => 'Killbill URL ex. http://127.0.0.1:8080'
+        method_option :database,
+                      :type    => :array,
+                      :default => nil,
+                      :desc    => 'DB connection parameters <database> <user> <password>'
+        desc 'account', 'export/import accounts'
+        def account
+          puts "Please wait processing the request!!!\n"
```
Comment (pierre) [inline kpm/lib/kpm/tasks.rb]: It would be good to have some progress displayed instead, maybe each time a new table is loaded for instance?

File: kpm/lib/kpm/database.rb [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            statements[table_name] << get_insert_statement(table_name,columns_names,value_data)
+          }
+        }
+
+        statements
+
+      end
+
+      def get_insert_statement(table_name, columns_names, values)
+        return "INSERT INTO #{table_name} ( #{columns_names} ) VALUES ( #{values}); SELECT LAST_INSERT_ID();"
```
Comment (pierre) [inline kpm/lib/kpm/database.rb]: Two thoughts:

* We would need a mechanism to display warnings too (see `SHOW WARNINGS` in the original script).
* I'm a bit concerned by the performance of generating single `INSERT` statements. We should at least do bulk loading. From https://dev.mysql.com/doc/refman/5.7/en/insert-optimization.html:

> If you are inserting many rows from the same client at the same time, use INSERT statements with multiple VALUES lists to insert several rows at a time. This is considerably faster (many times faster in some cases) than using separate single-row INSERT statements.

e.g. `INSERT INTO yourtable VALUES (1,2), (5,5), ...;`.

See also tips from https://dev.mysql.com/doc/refman/5.7/en/optimizing-innodb-bulk-data-loading.html .

File: kpm/lib/kpm/tasks.rb [hunk 8]
```diff
@@ -432,6 +432,97 @@ def system
...
+            if options[:killbill_api_credentials] && options[:killbill_api_credentials].size != 2
+              raise Interrupt,"\e[91;1m--killbill_api_credentials, required format -> <api_key> <api_secrets>\e[0m\n\n"
+            end
+
+            if options[:killbill_credentials] && options[:killbill_credentials].size != 2
+              raise Interrupt,"\e[91;1m--killbill_credentials, required format -> <user> <password>\e[0m\n\n"
+            end
+
+            if options[:database] && options[:database].size != 3
+              raise Interrupt,"\e[91;1m--database, required format -> <database> <user> <password>\e[0m\n\n"
```
Comment (pierre) [inline kpm/lib/kpm/tasks.rb]: IMHO, the format is a bit weird / unexpected:

```
kpm > ./bin/kpm account --import kbdump --database 'killbill killbill killbill'
Please wait processing the request!!!
--database, required format -> <database> <user> <password>
kpm > ./bin/kpm account --import kbdump --database killbill killbill killbill
Please wait processing the request!!!
Generating statements
```

I would implement individual switches instead (this would allow us to override only the database name for instance and keep the defaults otherwise). Additionally, support for `ENV` variables would be nice.

File: kpm/lib/kpm/account.rb [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+        puts "\e[91;1mAccount id not found\e[0m\n\n"
+        Dir.rmdir(TMP_DIR)
+      else
+        puts "\e[32mData exported under #{export_file}\e[0m\n\n"
+      end
+
+      export_file
+    end
+
+    def sanitize_export(table_name,col_name,value)
```
Comment (sbrossie) [inline kpm/lib/kpm/account.rb]: Nit: This is not really a sanity method but more an obfuscation, so for clarity you could change the name.

File: kpm/lib/kpm/account.rb [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+      if not File.exist?(source_file)
+        puts "\e[91;1mNeed to specify a valid file\e[0m\n\n"
+        Dir.rmdir(TMP_DIR)
+        return
+      end
+
+      sanitize_and_import(source_file, skip_payment_methods)
+    end
+
+    def sanitize_and_import(source_file, skip_payment_methods)
```
Comment (sbrossie) [inline kpm/lib/kpm/account.rb]: It would be nice if this method checks that the same number of rows that we had in the input for each table actually end up being inserted in the target table (maybe this already comes for free from the mysql import tool, unsure but something to check).

File: kpm/lib/kpm/account.rb [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+        return 1
+      elsif value.to_s === 'false'
+        return 0
+      else
+        return value
+      end
+    end
+
+    def fill_empty_column(value)
+      if value.to_s.strip.empty?
```
Comment (sbrossie) [inline kpm/lib/kpm/account.rb]: I have the feeling that we [used to insert](https://github.com/killbill/killbill/blob/master/bin/import-account#L52) `\N` in this case but maybe this is not required anymore because we are importing things differently?

File: kpm/lib/kpm/account.rb [hunk 12]
```diff
@@ -34,352 +35,412 @@ class Account
     REMOVE_DATA_FROM = {:accounts => [:name, :address1, :address2, :city, :state_or_providence, :phone, :email],
                        :account_history => [:name, :address1, :address2, :city, :state_or_providence, :phone, :email]}
 
-    def initialize(config_file = nil, killbill_api_credentials = nil, killbill_credentials = nil, killbill_url = nil, database = nil)
+    DATE_COLUMNS_TO_FIX = ['created_date','updated_date','processing_available_date']
```
Comment (pierre) [inline kpm/lib/kpm/account.rb]: There are probably more (e.g. `effective_date`). See http://docs.killbill.io/0.18/ddl.sql.

File: kpm/lib/kpm/database.rb [hunk 13]
```diff
@@ -4,72 +4,96 @@ module KPM
...
+      LAST_INSERTED_ID = 'SELECT LAST_INSERT_ID();'
+      ROWS_UPDATED = 'SELECT ROW_COUNT();'
+
       # Destination database
-      DATABASE = 'killbill'
-      USERNAME = 'root'
-      PASSWORD = 'root'
+      DATABASE = ENV['DATABASE'] || 'killbill'
+      USERNAME = ENV['USERNAME'] || 'root'
+      PASSWORD = ENV['PASSWORD'] || 'welcome0'
```
Comment (pierre) [inline kpm/lib/kpm/database.rb]: Use `root` as the default (convention we use elsewhere).

File: kpm/spec/kpm/unit/account_spec.rb [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+      "999999|#{dummy_account_id}|#{dummy_account_id}|willharnet@example.com|Will Harnet||USD|0||||UTC||||||||||||false|2017-04-03T15:50:14.000+0000|demo|2017-04-05T15:01:39.000+0000|Killbill::Stripe::PaymentPlugin|2\n"\
+      "-- account_history record_id|id|target_record_id|external_key|email|name|first_name_length|currency|billing_cycle_day_local|parent_account_id|payment_method_id|is_payment_delegated_to_parent|time_zone|locale|address1|address2|company_name|city|state_or_province|country|postal_code|phone|notes|migrated|is_notified_for_invoices|change_type|created_by|created_date|updated_by|updated_date|tenant_record_id\n"\
+      "999998|#{SecureRandom.uuid}|999999|#{dummy_account_id}|willharnet@example.com|Will Harnet||USD|0||||UTC||||||||||||false|INSERT|demo|2017-04-03T15:50:14.000+0000|demo|2017-04-03T15:50:14.000+0000|2\n"
+    }
+    let(:cols_names) {dummy_data.split("\n")[0].split(" ")[2]}
+    let(:cols_data) {dummy_data.split("\n")[1]}
+    let(:table_name) {dummy_data.split("\n")[0].split(" ")[1]}
+    let(:obfuscating_marker) {:email}
+
+    let(:mysql_cli) {"mysql #{db_name} --user=#{db_username} --password=#{db_password} -e"}
```
Comment (pierre) [inline kpm/spec/kpm/unit/account_spec.rb]: This test requires MySQL to be running, which is not always the case (e.g. on Travis). Could you create a new RSpec group (e.g. "test:mysql:spec") in the Rakefile so they are not run by default?

File: kpm/spec/kpm/unit/account_spec.rb [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+    logger.level = Logger::FATAL
+    logger}
+    let(:dummy_data_file) {Dir.mktmpdir('dummy') + File::SEPARATOR + 'kbdump'}
+    let(:url) {'http://127.0.0.1:8080'}
+    let(:killbill_api_key) {'bob'}
+    let(:killbill_api_secrets) {'lazar'}
+    let(:killbill_user) {'admin'}
+    let(:killbill_password) {'password'}
+    let(:db_name) {'killbill'}
+    let(:db_username) {'root'}
```
Comment (pierre) [inline kpm/spec/kpm/unit/account_spec.rb]: Quite risky as a default! 😸 

Could you instead use a specific user and database (e.g. 'test_kpm') and update the `README` to notify developers about the setup required to run these tests (database, credentials, etc.)?

File: kpm/spec/kpm/unit/account_spec.rb [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+      "-- account_history record_id|id|target_record_id|external_key|email|name|first_name_length|currency|billing_cycle_day_local|parent_account_id|payment_method_id|is_payment_delegated_to_parent|time_zone|locale|address1|address2|company_name|city|state_or_province|country|postal_code|phone|notes|migrated|is_notified_for_invoices|change_type|created_by|created_date|updated_by|updated_date|tenant_record_id\n"\
+      "999998|#{SecureRandom.uuid}|999999|#{dummy_account_id}|willharnet@example.com|Will Harnet||USD|0||||UTC||||||||||||false|INSERT|demo|2017-04-03T15:50:14.000+0000|demo|2017-04-03T15:50:14.000+0000|2\n"
+    }
+    let(:cols_names) {dummy_data.split("\n")[0].split(" ")[2]}
+    let(:cols_data) {dummy_data.split("\n")[1]}
+    let(:table_name) {dummy_data.split("\n")[0].split(" ")[1]}
+    let(:obfuscating_marker) {:email}
+
+    let(:mysql_cli) {"mysql #{db_name} --user=#{db_username} --password=#{db_password} -e"}
+    let(:mysql_delete_dummy_account) {"DELETE FROM accounts WHERE id = '#{dummy_account_id}'; SELECT ROW_COUNT();"}
```
Comment (pierre) [inline kpm/spec/kpm/unit/account_spec.rb]: Could you create the needed schema in a setup method?

File: kpm/spec/kpm/unit/account_spec.rb [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+
+    context 'when fetching account from api' do
+
+      it 'when account id not found' do
+        expect{ account_class.send(:fetch_export_data, account_id_invalid) }.to raise_error(Interrupt, 'Account id not found')
+      end
+
+      it 'when account id found' do
+        expect{ account_class.send(:fetch_export_data, account_id) }.not_to raise_error(Interrupt, 'Account id not found')
+        expect(account_class.send(:fetch_export_data, account_id)).to match(account_id)
```
Comment (pierre) [inline kpm/spec/kpm/unit/account_spec.rb]: This won't work unless the account exists in Kill Bill, right? If so, could you create it using the killbill-client-ruby library?

File: kpm/spec/kpm/unit/account_spec.rb [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+
+      it 'when round trip' do
+        open (dummy_data_file), 'w' do |io|
+          io.puts(dummy_data)
+        end
+        expect{account_class.import_data(dummy_data_file,false,10,true,true) }.not_to raise_error(Interrupt)
+        new_account_id = account_class.instance_variable_get(:@tables_id)
+
+        response = `#{mysql_cli} "DELETE FROM accounts WHERE id = '#{new_account_id['accounts_id']}'; SELECT ROW_COUNT();" 2>&1`
+        expect(response.split("\n")[1]).to eq('1')
```
Comment (pierre) [inline kpm/spec/kpm/unit/account_spec.rb]: I think you need to pass the password as `MYSQL_PWD` (see https://dev.mysql.com/doc/refman/5.5/en/password-security-user.html) instead of the command line.

On my setup, this assertion fails because of the leading warning message:

```
"mysql: [Warning] Using a password on the command line interface can be insecure.\nROW_COUNT()\n1\n"
```

File: kpm/lib/kpm/database.rb [hunk 19]
```diff
@@ -34,16 +38,21 @@ def set_database_name(database_name = nil)
...
+      def execute_insert_statement(table_name, query, qty_to_insert, table_data, record_id = nil)
 
         if not record_id.nil?
           query = "set #{record_id[:variable]}=#{record_id[:value]}; #{query}"
         end
         query = "SET autocommit=0; #{query} COMMIT;"
-        response = `#{@@mysql_command_line} "#{query.gsub('"','\\"')}" 2>&1`
+        
+        File.open(STATEMENT_TMP_FILE,'w') do |s|
+          s.puts query.gsub('"','\\"')
```
Comment (pierre) [inline kpm/lib/kpm/database.rb]: Single quotes need to be escaped as well: I just had the issue with a `reason` field containing the word `DIDN'T`. Worth a unit test maybe?

**Comment by pierre** [comment]
Thanks! Since the conversation is getting hard to follow, I've summarized the remaining items to address in https://github.com/killbill/killbill-cloud/issues/83.
